### ISSUE TITLE
Concurrency issues in GetImplementations caching mechanism

### ISSUE THREAD

**Issue by PTwr** [ExtCore/ExtCore#123]
Multiple calls to GetImplementations from bunch to background tasks running on app start leads to `System.ArgumentException: 'An item with the same key has already been added.` when `useCaching = true`.

Between querying cache and filling it with obtained types other threads seem to finish first so `Add()` on dict fails with exception.

Switching to `ConcurentDictionary` might take care of this issue, something like this:

https://github.com/PTwr/ExtCore/commit/5c4096936f5c2118b379da2a5d15990b16f0e791#diff-4145a0ed1b7433005dde69197e020266

however some lock&wait might be in order to avoid molesting all Assemblies needlessly.

Another possible solution might be prefetching `GetExportedTypes()` on initialization, for example in `UseExtCore()` with option to reload/append to it in case #43 is implemented.

**Comment by DmitrySikorsky** [comment]
Thank you for your detailed issue. I will fix it ASAP.

**Comment by PTwr** [comment]
I'll try to find time this week to change `ExtensionManager` from `static` to DI registered as `Singleton` as my current project would enjoy custom implementation of `ExtensionManager` and #43 (even limited). Adding prefetching of Assemblies and/or ExportedTypes would be easy to add during those changes.

Would such changes fit your idea of ExtCore? It would be convenient for my team if it could be merged to mainline release :)

**Comment by DmitrySikorsky** [comment]
Bug is fixed (using your solution).

What about using ExtensionManager as a service. Hm. The only issue I see is that we use on early stages, so not all the services can be registered when you request it from the ServiceProvider. For example,  here: https://github.com/ExtCore/ExtCore/blob/master/src/ExtCore.Data/Actions/AddStorageAction.cs#L37

What do you think about that?

**Comment by PTwr** [comment]
Sorry for late answer, evil real life attacked. Thanks for fix, I was about to profile performance loss  due to no caching :)

I had indeed thought about that problem. Due to strictness of C# possible solution is quite ugly, a dummy static class with singleton for implementation: https://gist.github.com/PTwr/be475a071e28d39df06c8ed564560f3a

default/base implementation sets itself as singleton value in constructor, so once overridden implementation is created it will replace singleton (eg. register own implementation `AsSingleton` and inject it in startup `Configure(IApplicationBuilder app...`)

it would work... but its horrible "solution" that would probably get me killed after code review, damn static classes always end up causing problems in future.

Better alternative might be registering callbacks/events in current static class, however new cache system should offer all I really need so I no longer have need to override it. Fancy stuff can now be done on my side. And we finally got partial implementation of #43 and #44 :)
